### ISSUE TITLE
Customer Attribute gets nulled

### ISSUE THREAD

**Issue by jakwinkler** [magento/magento2#25370]
### Preconditions (*)

1. Magento 2.3.3 & 2.3.2 Community Edition

### Steps to reproduce (*)

1. Add **2** new customer attributes via InstallData script
(part of the code)
```
$eavSetup->addAttribute(
            Customer::ENTITY,
            'custom_attribute',
            [
                'type'         => 'int',
                'label'        => 'Customer custom attribute nr 1',
                'input'        => 'text',
                'required'     => false,
                'visible'      => true,
                'user_defined' => true,
                'position'     => 1000,
                'system'       => false,
            ]
        );
```

2. Add a new block to customer dashboard account information via XML to this reference container `<referenceContainer name="form.additional.info">`
3. Assign that block a custom template file with new form element that relates to one of customer attribute, for example
```
<select id="something" class="input__field" name="binary_branch">
        <option value="1"><?= __('Dark side); ?></option>
        <option value="2"><?= __('Jedi Order'); ?</option>
</select>
```
4. Add customer attributes ui_components so they can be visible in the backend
5. Create customer account
6. Modify custom attributes in backend, assign some random data
7. Go to Account Edit in Customer Dashboard
8. Change value of custom attrbute
9. Save customer 

### Expected result (*)

1. Customer attribute defined in additional form element gets updated

### Actual result (*)

1. First customer attribute gets updated properly
2. Second customer attribute gets nulled

**Comment by kanthy-magento** [comment]
@magento give me 2.3-develop instance

**Comment by magento-engcom-team** [comment]
Hi @itsPranith. Thank you for your request. I'm working on Magento 2.3-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @itsPranith, here is your [Magento instance](https://i-25370-2-3-develop.instances.magento-community.engineering).
Admin access: [https://i-25370-2-3-develop.instances.magento-community.engineering/admin](https://i-25370-2-3-develop.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @itsPranith 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-22576` were created

**Issue Available**: @itsPranith, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by rhoerr** [comment]
I believe I've been tracking the same issue. I haven't gone to older versions to verify, but I haven't seen any obvious code changes since 2.2.0 that would explain the behavior, so I suspect it's always been this way.

-----

Rough notes on what's going on, to help whoever investigates this: It has to do with hidden attributes (any not appearing on the frontend customer form) being interpreted as `false` by the customer form extractor, and that false value being propagated to save.

The data disappears in `$this->customerExtractor->extract(...)` in EditPost controller.

That calls `$customerForm->extractData($request)` and then `$customerForm->compactData($data)`. `extractData` gives `false` for any attributes that aren't in the request, or are empty in it _(*at least for date and select attribute types--I didn't check all)_. Those methods are here: https://github.com/magento/magento2/blob/2.3.3/app/code/Magento/Customer/Model/Metadata/Form.php#L235

They call `dataModel->extractValue` and `->compactValue`. `extractValue` ultimately calls https://github.com/magento/magento2/blob/2.3.3/app/code/Magento/Eav/Model/Attribute/Data/AbstractData.php#L512 -- Note that it defaults to `false` if it can't find a value in the request--which it won't, because they aren't part of the form or posted data, and shouldn't be.

And then the `compactValue()` methods for most or all attribute types just return `$value` -- and so false gets set on the data object that becomes the new saved customer. Ergo, the hidden attr values get erased on frontend save.

-----

This is probably not a generally acceptable solution and I'm sure there are unintended consequences, so *do this at your own risk*:

I worked around the bug by overwriting (preferencing) `\Magento\Customer\Model\Metadata\Form`, specifically changing `compactData()` to skip false attributes (retaining the existing value) instead of persisting them.

```
    /**
     * Compact data array to form attribute values
     *
     * @param array $data
     * @return array attribute values
     */
    public function compactData(array $data)
    {
        foreach ($this->getAllowedAttributes() as $attribute) {
            $dataModel = $this->_getAttributeDataModel($attribute);
            $dataModel->setExtractedData($data);
            $attributeCode = $attribute->getAttributeCode();
            /**
             * CORE BUG FIX: Frontend account edit save will result in any hidden attributes getting value "false"
             * (because they don't exist in the request), and this persists those falses onto the saved model.
             * Result: Hidden attribute values all get erased on save. Bad. This skips them instead so the original
             * values persist.
             * @see https://github.com/magento/magento2/issues/25370
             */
            if (!in_array($attributeCode, ['created_at', 'created_in'], true)
                && (!isset($data[$attributeCode]) || $data[$attributeCode] === false)) {
                // $data[$attribute->getAttributeCode()] = false;
                continue;
            }
            $this->_attributeValues[$attributeCode] = $dataModel->compactValue($data[$attributeCode]);
        }

        return $this->_attributeValues;
    }
```

I would much rather have done this in an after plugin on `EditPost::populateNewCustomerDataObject()`, but that method (and everything else around it) is private and untouchable short of rewriting the entire EditPost controller.

**Comment by qsolutions-pl** [comment]
Is anyone working on this bug?

**Comment by ralbin** [comment]
@magento I am working on this

**Comment by ralbin** [comment]
@magento give me 2.4-develop instance

**Comment by magento-engcom-team** [comment]
Hi @ralbin. Thank you for your request. I'm working on Magento 2.4-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @ralbin, here is your [Magento instance](https://i-25370-2-4-develop.instances.magento-community.engineering).
Admin access: [https://i-25370-2-4-develop.instances.magento-community.engineering/admin_d082](https://i-25370-2-4-develop.instances.magento-community.engineering/admin_d082)
Login: `4b59444a` Password: `efdd15202900`
Instance will be terminated in up to 3 hours.

**Comment by ihorvansach** [comment]
Today I have faced with the similar issue on one Magento instance with the custom customer attribute. Attribute was visible/editable in the admin panel only (Admin panel customer edit page), but once customer saved it's account on the storefront this attribute value becomes empty.

Issue was related to wrong attribute settings. Attribute was assigned to these forms: adminhtml_checkout, adminhtml_customer, checkout_register, customer_account_create, customer_account_edit, so Magento thought that it should be on the customer edit page on the storefront as well.

What I did?
1. I run this SQL query to get all current forms assigned to the attribute:
```sql
SELECT * FROM `customer_form_attribute` WHERE `attribute_id` IN (
    SELECT attribute_id FROM eav_attribute
        WHERE attribute_code="YOUR_ATTRIBUTE_CODE_HERE" 
)
````
And then removed all except adminhtml_customer (as the attribute in my case should be only on the customer editing form in the admin panel ).

2. As this attribute came to the Magento 2 from the Magento 1 (during the migration), I have also created a simple extension to re-setup the attribute (used mage2gen.com to quickly generate the module).

So basically it was wrong attribute setup.

@qsolutions-pl , @rhoerr , please try to re-check an attribute form settings and set proper values. Also if an attribute was created automatically during the migration, and was not defined via setup or setup-patch script, try to reinstall it.

**Comment by qsolutions-pl** [comment]
It has been 2 years and none-work has been put into this nor it was solved. 
I'm closing the issue.
